src/relations/backend_database.py

```diff
@@ -69,45 +81,152 @@ def __init__(self, charm: CharmBase):
+        )
+        cfg = self.charm.read_pgb_config()
+        # adds user to pgb config
+        cfg.add_user(user=event.username, admin=True)
+        cfg["pgbouncer"][
+            "auth_query"
+        ] = f"SELECT username, password FROM {self.auth_user}.get_auth($1)"
+        cfg["pgbouncer"]["auth_file"] = f"{PGB_DIR}/userlist.txt"
+        self.charm.render_pgb_config(cfg)
+
+        self.charm.update_postgres_endpoints(reload_pgbouncer=True)
+
+    def _on_endpoints_changed(
+        self, _: Union[DatabaseEndpointsChangedEvent, DatabaseReadOnlyEndpointsChangedEvent]
+    ):
+        self.charm.update_postgres_endpoints(reload_pgbouncer=True)
+
+    def _on_relation_departed(self, event: RelationDepartedEvent):
+        if event.departing_unit != self.charm.unit:
+            return
+
+        logger.info("removing auth user")
+
+        uninstall_script = open("src/relations/pgbouncer-uninstall.sql", "r").read()
+
+        with self.postgres.connect_to_database(PGB_DB) as conn, conn.cursor() as cursor:
+            cursor.execute(uninstall_script.replace("auth_user", self.auth_user))
+        conn.close()
 
-    def _on_relation_broken(self, event: RelationBrokenEvent):
+        logger.info("auth user removed")
+
+    def _on_relation_broken(self, _: RelationBrokenEvent):
         """Handle backend-database-relation-broken event.
 
         Removes all traces of this relation from pgbouncer config.
         """
-        try:
-            cfg = self.charm.read_pgb_config()
-            user = self.charm.backend_postgres.user
-            self.charm.remove_user(user, cfg=cfg, reload_pgbouncer=True, render_cfg=True)
+        cfg = self.charm.read_pgb_config()
+        cfg.remove_user(self.postgres.user)
+        cfg["pgbouncer"].pop("auth_user", None)
+        cfg["pgbouncer"].pop("auth_query", None)
+        self.charm.render_pgb_config(cfg)
+        self.charm.delete_file(f"{PGB_DIR}/userlist.txt")
 
-            self.charm.update_postgres_endpoints()
+        self.charm.remove_postgres_endpoints(reload_pgbouncer=True)
 
-        except FileNotFoundError:
-            logging.error("failed to access config file")
-            event.defer()
+    def initialise_auth_function(self, dbname=PGB_DB):
+        """Runs an SQL script to initialise the auth function.
+
+        This function must run in every database for authentication to work.
+
+        Args:
+            postgres: a PostgreSQL application instance. If none is provided, the default pgbouncer
+                instance will be used.
+            dbname: the name of the database to connect to.
+        """
+        logger.info("initialising auth function")
+
+        install_script = open("src/relations/pgbouncer-install.sql", "r").read()
+
+        with self.postgres.connect_to_database(dbname) as conn, conn.cursor() as cursor:
+            cursor.execute(install_script.replace("auth_user", self.auth_user))
```

Comment:
is there any error checking that should be done?

Reply:
This function is only being used in relation_joined hooks, where we're deferring if self.postgres isn't accessible. Running this basically anywhere else is a bad idea, so I'm happy to let this error out, but I'll document how and why it works.